File: src/main/java/org/cubewhy/celestial/gui/pages/GuiPlugins.java
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */

/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */

package org.cubewhy.celestial.gui.pages;

import lombok.SneakyThrows;
import org.cubewhy.celestial.files.DownloadManager;
import org.cubewhy.celestial.files.Downloadable;
import org.cubewhy.celestial.game.RemoteAddon;
import org.cubewhy.celestial.game.addon.JavaAgent;
import org.cubewhy.celestial.game.addon.LunarCNMod;
import org.cubewhy.celestial.game.addon.WeaveMod;
import org.cubewhy.celestial.gui.layouts.VerticalFlowLayout;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.border.TitledBorder;
import java.awt.*;
import java.io.File;
import java.net.URL;
import java.util.List;

import static org.cubewhy.celestial.Celestial.f;
import static org.cubewhy.celestial.Celestial.launcherData;

public class GuiPlugins extends JPanel {
    private final JTabbedPane tab;

    public GuiPlugins() {
        this.setBorder(new TitledBorder(null, f.getString("gui.plugin.title"), TitledBorder.DEFAULT_JUSTIFICATION, TitledBorder.DEFAULT_POSITION, null, Color.orange));
        this.setLayout(new BoxLayout(this, BoxLayout.Y_AXIS));

        this.tab = new JTabbedPane();

        this.initGui();
    }

    private void initGui() {
        this.add(tab);
        addTabs();

        // refresh addons
        JButton btnRefresh = new JButton(f.getString("gui.plugin.refresh"));
        btnRefresh.addActionListener((e) -> {
            tab.removeAll();
            addTabs();
        });
        this.add(tab);
        this.add(btnRefresh);
    }

    @SneakyThrows
    private void addTabs() {
        List<RemoteAddon> addons = launcherData.getPlugins();
        if (addons == null) {
            this.add(new JLabel(f.getString("gui.plugin.unsupported")));
            return;
        }
        JPanel panelWeave = new JPanel();
        panelWeave.setLayout(new VerticalFlowLayout());
        JPanel panelAgents = new JPanel();
        panelAgents.setLayout(new VerticalFlowLayout());
        JPanel panelCN = new JPanel();
        panelCN.setLayout(new VerticalFlowLayout());
        JScrollPane sp1 = new JScrollPane(panelWeave, ScrollPaneConstants.VERTICAL_SCROLLBAR_AS_NEEDED, ScrollPaneConstants.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        sp1.getVerticalScrollBar().setUnitIncrement(30);
        JScrollPane sp2 = new JScrollPane(panelAgents, ScrollPaneConstants.VERTICAL_SCROLLBAR_AS_NEEDED, ScrollPaneConstants.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        sp2.getVerticalScrollBar().setUnitIncrement(30);
        JScrollPane sp3 = new JScrollPane(panelCN, ScrollPaneConstants.VERTICAL_SCROLLBAR_AS_NEEDED, ScrollPaneConstants.HORIZONTAL_SCROLLBAR_AS_NEEDED);
        sp3.getVerticalScrollBar().setUnitIncrement(30);
        this.tab.addTab("Weave", sp1);
        this.tab.addTab("Agents", sp2);
        this.tab.addTab("LunarCN", sp3);
        for (RemoteAddon addon : addons) {
            if (addon.getCategory() == RemoteAddon.Category.WEAVE) {
                addPlugin(panelWeave, addon, WeaveMod.modFolder);
            } else if (addon.getCategory() == RemoteAddon.Category.AGENT) {
                addPlugin(panelAgents, addon, JavaAgent.javaAgentFolder);
            } else if (addon.getCategory() == RemoteAddon.Category.CN) {
                addPlugin(panelCN, addon, LunarCNMod.modFolder);
            }
        }
    }

    /**
     * Add label and download button
     *
     * @param panel  the panel
     * @param addon  the addon
     * @param folder target folder
     */
    private void addPlugin(JPanel panel, @NotNull RemoteAddon addon, File folder) {
        JPanel p = new JPanel();
        p.setLayout(new GridLayout());
        p.add(new JLabel(addon.getName()));
        File file = new File(folder, addon.getName());
        if (file.exists()) {
            p.add(new JLabel(f.getString("gui.plugin.exist")));
        } else {
            p.add(getDownloadButton(addon.getDownloadURL(), file));
        }
        panel.add(p);
    }

    private @NotNull JButton getDownloadButton(URL url, @NotNull File file) {
        JButton btn = new JButton(String.format(f.getString("gui.plugins.download"), file.getName()));
        btn.addActionListener((e) -> {
            DownloadManager.download(new Downloadable(url, file, null));
            new Thread(() -> {
                try {
                    DownloadManager.waitForAll();
                } catch (InterruptedException err) {
                    throw new RuntimeException(err);
                }
            }).start();
        });
        return btn;
    }
}

File: src/main/kotlin/org/cubewhy/celestial/gui/pages/GuiPlugins.kt
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */
package org.cubewhy.celestial.gui.pages

import org.cubewhy.celestial.f
import org.cubewhy.celestial.launcherData
import org.cubewhy.celestial.format
import org.cubewhy.celestial.game.RemoteAddon
import org.cubewhy.celestial.game.addon.JavaAgent
import org.cubewhy.celestial.game.addon.LunarCNMod
import org.cubewhy.celestial.game.addon.WeaveMod
import org.cubewhy.celestial.gui.dialogs.AddonInfoDialog
import org.cubewhy.celestial.gui.layouts.VerticalFlowLayout
import org.cubewhy.celestial.withScroller
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.Color
import java.awt.GridLayout
import java.io.File
import javax.swing.*
import javax.swing.border.TitledBorder

class GuiPlugins : JPanel() {
    private val tab: JTabbedPane

    companion object {
        private val log: Logger = LoggerFactory.getLogger(GuiPlugins::class.java)
    }

    init {
        this.border = TitledBorder(
            null,
            f.getString("gui.plugins.title"),
            TitledBorder.DEFAULT_JUSTIFICATION,
            TitledBorder.DEFAULT_POSITION,
            null,
            Color.orange
        )
        this.layout = BoxLayout(this, BoxLayout.Y_AXIS)

        this.tab = JTabbedPane()

        this.initGui()
    }

    private fun initGui() {
        this.add(tab)
        addTabs()

        // refresh addons
        val btnRefresh = JButton(f.getString("gui.plugins.refresh"))
        btnRefresh.addActionListener {
            this.refresh()
        }
        this.add(tab)
        this.add(btnRefresh)
    }

    private fun refresh() {
        this.tab.removeAll()
        this.addTabs()
    }

    private fun addTabs() {
        val addons: List<RemoteAddon>? = launcherData.plugins
        if (addons == null) {
            this.add(JLabel(f.getString("gui.plugins.unsupported")))
            return
        }
        val panelWeave = JPanel()
        panelWeave.layout = VerticalFlowLayout()
        val panelAgents = JPanel()
        panelAgents.layout = VerticalFlowLayout()
        val panelCN = JPanel()
        panelCN.layout = VerticalFlowLayout()
        tab.addTab("Weave", panelWeave.withScroller())
        tab.addTab("Agents", panelAgents.withScroller())
        tab.addTab("LunarCN", panelCN.withScroller())
        for (addon in addons) {
            when (addon.category) {
                RemoteAddon.Category.WEAVE -> {
                    addPlugin(panelWeave, addon, WeaveMod.modFolder)
                }

                RemoteAddon.Category.AGENT -> {
                    addPlugin(panelAgents, addon, JavaAgent.javaAgentFolder)
                }

                RemoteAddon.Category.CN -> {
                    addPlugin(panelCN, addon, LunarCNMod.modFolder)
                }
            }
        }
    }

    /**
     * Add label and download button
     *
     * @param panel  the panel
     * @param addon  the addon
     * @param folder target folder
     */
    private fun addPlugin(panel: JPanel, addon: RemoteAddon, folder: File) {
        val p = JPanel()
        p.layout = GridLayout()
        p.add(JLabel(addon.name))
        val file = File(folder, addon.name)
        p.add(getInfoButton(addon, file))
        panel.add(p)
    }

    private fun getInfoButton(addon: RemoteAddon, file: File): JButton {
        val btn = JButton(f.format("gui.plugins.info", file.name))
        btn.addActionListener {
            log.info("Open plugin info dialog for " + addon.name)
            AddonInfoDialog(addon, file).isVisible = true // show info dialog
        }
        return btn
    }
}
